File: DmModule/python/DmModule/Parameters.py
#
# Copyright (C) 2012-2020 Euclid Science Ground Segment
#
# This library is free software; you can redistribute it and/or modify it under
# the terms of the GNU Lesser General Public License as published by the Free
# Software Foundation; either version 3.0 of the License, or (at your option)
# any later version.
#
# This library is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this library; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
#


"""
File: python/DmModule/Parameters.py

Created on: 09/24/20
Author: user
"""

import argparse
import ElementsKernel.Logging as log

#==============================================================================================
# EXERCISE
#==============================================================================================
# Include the python packages and modules to read parameters from XML file.

from ST_DataModelBindings.dpd.le3.wl.twodmass.inp.raw.paramsconvergencepatch_stub as param

class Parameters:

  def readParameterFile (parameter_file):
        logger = log.getLogger('Parameters')
        logger.info('Getting information from input parameter XML file ' + parameter_file + '...')
        
        # Parse the XML file and create the binding object
        logger.debug('Parsing file ' + parameter_file + '...')

        #
        # Exercise
        # ------------------------------------------------------
        # Parse the input parameter_file in a binding object

        logger.info('TO DO: Parse the input parameter_file in a binding object')
        #with open(parameter_file, "r") as f:
        #        Param_xml_string = f.read()
        #Param_xml = param.CreateFromDocument(Param_xml_string)
        #logger.debug('Binding object created successfully')

        #
        # Exercise
        # ------------------------------------------------------
        # Get and log the Parameters from the input parameter file

        logger.info('TO DO: Get and log the Parameters from the input parameter file')
        #self.m_NInpaint = Param_xml.Data.GapsParams.NInpaint
        #logger.info('Number of Inpainting iterations : ' + str(self.m_NInpaint))
        #self.m_NItReducedShear = Param_xml.Data.ReducedShear.NItReducedShear
        #logger.info('Number of iterations for Reduced shear : ' + str(self.m_NItReducedShear))
        #self.m_RSsigmaGauss = Param_xml.Data.ReducedShear.GaussSTD

        #
        # Exercise
        # ------------------------------------------------------
        # Check if the optional element ThresholdFDR exists in the XML

        logger.info('TO DO: Check if the optional element ThresholdFDR exists in the XML')
        #if Param_xml.Data.ReducedShear.ThresholdFDR:
        #logger.debug('The XML file contains ThresholdFDR for reduced shear')
        #self.m_RSthresholdFDR = Param_xml.Data.ReducedShear.ThresholdFDR
        #logger.info('ThresholdFDR for reduced shear : ' + str(self.m_RSthresholdFDR))

        # ------------------------------------------------------
        # ------------------------------------------------------
        logger.info('TO DO: Log the values of the other elements (optional/iterable parameters)')
        # ------------------------------------------------------
        # ------------------------------------------------------
        self.m_nbZBins = Param_xml.Data.RedshiftBins.Nbins
        logger.info('m_nbZBins: ' + str(self.m_nbZBins))
        self.m_balancedBin = Param_xml.Data.RedshiftBins.BalancedBins
        logger.info('m_balancedBin: ' + str(self.m_balancedBin))
        for redshift_bin in Param_xml.Data.RedshiftBins.RedshiftBin:
            self.m_zMax = redshift_bin.ZMax
            self.m_zMin = redshift_bin.ZMin
            logger.info('m_zMax : ' + str(self.m_zMax))
            logger.info('m_zMin : ' + str(self.m_zMin))
        # sometimes parameter can appear more than one time, its corresponding method returns a sequence
        # Iterate over individual records.
        #
        # ------------------------------------------------------
        # Log the values of the elements
        #
        # Tip: The special sequence type is iterable
        #
        for patches in Param_xml.Data.PatchParams:
            self.m_nbPatches = patches.NPatches
            logger.info('m_nbPatches : ' + str(self.m_nbPatches))
            # If an element can appear more than one times, it is represented as a list
            for patchlist in patches.PatchList:
                  self.m_PatchWidth = patchlist.PatchWidth
                  logger.info('PatchWidth : ' + str(self.m_PatchWidth))
                  self.m_PixelSize = patchlist.PixelSize
                  logger.info('PixelSize : ' + str(self.m_PixelSize/60.))
                  self.CenterX = patchlist.ProjCtr.Longitude
                  self.CenterY = patchlist.ProjCtr.Latitude
                  logger.info('CenterX : ' + str(self.CenterX))
                  logger.info('CenterY : ' + str(self.CenterY))

        # ------------------------------------------------------
        # ------------------------------------------------------

  def __init__(self):
          """
          The entry point to create an an Fits keywords object.
          """
          self.m_NInpaint = 100
          self.m_NItReducedShear = 10
          self.m_nbPatches = 1
          self.m_PixelSize = 0.586/60.
          self.m_PatchWidth = 10
          self.mapCenterX = []
          self.mapCenterY = []
          self.m_nbZBins = 1
          self.m_zMin = []
          self.m_zMax = 10
          self.m_balancedBin = 0
          self.m_EqualVarPerScale = 0
          self.m_ForceBMode = 0
          self.m_nbScales = 0
          self.m_add_borders = 0
          self.m_RSsigmaGauss = 0.
          self.m_sigmaGauss = 0.
          self.m_nbSamples = 0
          self.m_RSthresholdFDR = 0
          self.m_thresholdFDR = 0

  def getNInpaint(self):
          return self.m_NInpaint

  def getZMin(self):
          return self.m_zMin

  def getZMax(self):
          return self.m_zMax

  def getPixelsize(self):
          return self.m_PixelSize

  def getSigmaGauss(self):
          return self.m_sigmaGauss

  def getThreshold(self):
          return self.m_thresholdFDR

  def getRSSigmaGauss(self):
          return self.m_RSsigmaGauss

  def getRSThreshold(self):
          return self.m_RSthresholdFDR

  def getnbZBins(self):
          return self.m_nbZBins

  def getnbPatches(self):
          return self.m_nbPatches

  def getNItReducedShear(self):
          return self.m_NItReducedShear

  def getnbScales(self):
          return self.m_nbScales

  def getPatchWidth(self):
          return self.m_PatchWidth

  def getMapCenterX(self):
          return self.mapCenterX

  def getMapCenterY(self):
          return self.mapCenterY

  def getNSamples(self):
          return self.m_nbSamples

  def get_addBorders(self):
          return self.m_add_borders

  def getEqualVarPerScale(self):
          return self.m_EqualVarPerScale

  def getForceBMode(self):
          return self.m_ForceBMode

  def get_BalancedBins(self):
          return self.m_balancedBin

File: DmModule/src/lib/Parameters.cpp
/**
 * @file src/lib/Parameters.cpp
 * @date 09/22/20
 * @author user
 *
 * @copyright (C) 2012-2020 Euclid Science Ground Segment
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 3.0 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 *
 */

#include "DmModule/Parameters.h"

static Elements::Logging logger = Elements::Logging::getLogger("Parameters");

namespace DmModule {

 Parameters::Parameters():m_NItReducedShear(10), m_nbPatches(1), m_PixelSize( 0.586/60.), m_PatchWidth(10.),
           mapCenterX(0.), mapCenterY(0.), m_nbZBins(1), m_zMin(0.), m_zMax(10.), m_balancedBin(0),
           m_NInpaint(100), m_EqualVarPerScale(0), m_ForceBMode(1), m_nbScales(0), m_add_borders(0),
           m_sigmaGauss(0.), m_thresholdFDR(0.), m_nbSamples(0), m_RSsigmaGauss(0.), m_RSthresholdFDR(0.)
 { }

 Parameters::Parameters(int NItReducedShear, int NPatches, float PixelSize, float PatchWidth, std::vector<double> mapCenterX,
          std::vector<double> mapCenterY, int nbZBins, std::vector<double> zMin, double zMax, long BalancedBins,
          int NInpaint, long EqualVarPerScale, long ForceBMode, int nbScales, long add_borders, float RSsigmaGauss,
          float sigmaGauss, int nbSamples, float RSthresholdFDR, float thresholdFDR): m_NItReducedShear(NItReducedShear), m_nbPatches(NPatches),
           m_PixelSize(PixelSize/60.), m_PatchWidth(10.), mapCenterX(mapCenterX), mapCenterY(mapCenterY),
           m_nbZBins(nbZBins), m_zMin(zMin), m_zMax(zMax), m_balancedBin(BalancedBins),
           m_NInpaint(NInpaint), m_EqualVarPerScale(EqualVarPerScale), m_ForceBMode(ForceBMode),
           m_nbScales(nbScales), m_add_borders(add_borders), m_RSsigmaGauss(RSsigmaGauss), m_sigmaGauss(sigmaGauss),
           m_thresholdFDR(thresholdFDR), m_RSthresholdFDR(RSthresholdFDR), m_nbSamples(nbSamples) { }

 Parameters Parameters::readParameterFile (const boost::filesystem::path& parameter_file){
  logger.info() << "Getting information from input Parameter XML file " << parameter_file << " ...";
  logger.debug() << "Parsing file " << parameter_file << " ...";

  // --------------------------------------------------------------
  // Exercise
  // --------------------------------------------------------------
  // Parse the file in_xml_filename in a binding object.

  logger.warn() << "TODO: Parse the file " << parameter_file
                << " in a binding object";
   //auto Param_xml = dpd::le3::wl::twodmass::inp::paramsconvergencepatch::DpdTwoDMassParamsConvergencePatch
   //                     (parameter_file, xml_schema::flags::dont_validate);

  // Exercise
  // -------------------------------------------------------
  // Get and log the Parameters from the input parameter file

  logger.warn() << "TODO: Get and log the Parameters from the input parameter file";

  // Exercise
  // -------------------------------------------------------
  // Check if the optional element ThresholdFDR exists in the XML
  logger.warn() << "TODO: Check if the ThresholdFDR element exists";

   if (true) {

    logger.debug() << "The ThresholdFDR element is present";

    logger.warn() << "TODO: Log the value of the ThresholdFDR element";

  }

    // sometimes parameter can appear more than one time, its corresponding method returns a sequence
   // Iterate over individual records.
    //
    //Exercise
    // ------------------------------------------------------
    // Log the values of the elements
    //
    // Tip: The special sequence type is iterable
    //
    logger.warn() << "TODO: Log the values of the elements";

 return Parameters(m_NItReducedShear, m_nbPatches, m_PixelSize, m_PatchWidth, mapCenterX, mapCenterY, m_nbZBins,
            m_zMin, m_zMax, m_balancedBin, m_NInpaint, m_EqualVarPerScale, m_ForceBMode, m_nbScales, m_add_borders,
            m_RSsigmaGauss, m_sigmaGauss, m_nbSamples, m_RSthresholdFDR, m_thresholdFDR);
 }

 std::vector<double> Parameters::getZMin(){ return m_zMin; }
 double Parameters::getZMax(){ return m_zMax; }
 float Parameters::getPixelsize(){ return m_PixelSize; }
 float Parameters::getSigmaGauss(){ return m_sigmaGauss; }
 float Parameters::getThreshold(){return m_thresholdFDR; }
 float Parameters::getRSSigmaGauss(){ return m_RSsigmaGauss; }
 float Parameters::getRSThreshold(){return m_RSthresholdFDR; }
 int Parameters::getnbZBins(){ return m_nbZBins; }
 int Parameters::getnbPatches(){ return m_nbPatches; }
 int Parameters::getNInpaint(){ return m_NInpaint; }
 int Parameters::getNItReducedShear(){ return m_NItReducedShear; }
 int Parameters::getnbScales(){ return m_nbScales; }
 float Parameters::getPatchWidth(){ return m_PatchWidth; }
 std::vector<double> Parameters::getMapCenterX(){ return mapCenterX; }
 std::vector<double> Parameters::getMapCenterY(){ return mapCenterY; }
 int Parameters::getNSamples() { return m_nbSamples;}

 long Parameters::get_addBorders(){
  return m_add_borders;
 }
 long Parameters::getEqualVarPerScale(){
  return m_EqualVarPerScale;
 }
 long Parameters::getForceBMode(){
  return m_ForceBMode;
 }
 long Parameters::get_BalancedBins(){
  return m_balancedBin;
 }

}  // namespace DmModule
